### PR TITLE
Preserve no break spaces in Latex to Unicode conversion

### LINKED ISSUE #15158: Convert LaTeX to Unicode feature should preseve no-break spaces
The converter currently replaces LaTeX no-break spaces (`~`) with Unicode plain spaces (` ` or "\u0020" ).
These should be converted as no-break spaces (` ` or "\u00a0").

### PR DESCRIPTION
**Pull request by faneeshh**
Closes #15158
PR Description

I updated the Latex to Unicode conversion to ensure the tilde (~) is converted to a non breaking space (\u00a0) instead of a standard space and I should mention that I used a negative lookbehind regex to make sure this only affects standalone tildes and doesn't break Latex tilde accents like \~{n}. I think this would maintain support for accented characters.

Screenshot

image

### Checklist

- [x] I own the copyright of the code submitted and I license it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] I manually tested my changes in running JabRef (always required)
- [x] I added JUnit tests for changes (if applicable)
- [/] I added screenshots in the PR description (if change is visible to the user)
- [x] I added a screenshot in the PR description showing a library with a single entry with me as author and as title the issue number
- [x] I described the change in `CHANGELOG.md` in a way that can be understood by the average user (if change is visible to the user)
- [x] I checked the [user documentation](https://docs.jabref.org/) for up to dateness and submitted a pull request to our [user documentation repository](https://github.com/JabRef/user-documentation/tree/main/en)

### CODE REVIEW

**Comment by aniruddhkorde9-dot** [comment]
Hello! I am a new contributor looking to get involved. I would love to work on this issue. Could you please assign it to me?

**Comment by subhramit** [comment]
> Hello! I am a new contributor looking to get involved. I would love to work on this issue. Could you please assign it to me?

Are you aware that this is a PR and not an issue?

File: jablib/src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java [hunk 1]
```diff
@@ -13,6 +13,10 @@ public class LatexToUnicodeAdapter {
 
     private static final Pattern UNDERSCORE_MATCHER = Pattern.compile("_(?!\\{)");
 
+    private static final Pattern TILDE_MATCHER = Pattern.compile("(?<!\\\\)~");
+
+    private static final String NON_BREAKING_SPACE = "\u00a0";
```
Comment (ryan-carpenter) [inline jablib/src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java]:
```suggestion
    private static final String NO_BREAK_SPACE = "\u00a0";
```

File: jablib/src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java [hunk 2]
```diff
@@ -30,7 +34,8 @@ public static String format(@NonNull String inField) {
     /// @param inField a String containing LaTeX
     /// @return an `Optional<String>` with LaTeX resolved into Unicode or `empty` on failure.
     public static Optional<String> parse(@NonNull String inField) {
-        String toFormat = UNDERSCORE_MATCHER.matcher(inField).replaceAll(REPLACEMENT_CHAR);
+        String toFormat = TILDE_MATCHER.matcher(inField).replaceAll(NON_BREAKING_SPACE);
```
Comment (ryan-carpenter) [inline jablib/src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java]:
```suggestion
        String toFormat = TILDE_MATCHER.matcher(inField).replaceAll(NO_BREAK_SPACE);
```

File: jablib/src/test/java/org/jabref/logic/layout/format/LatexToUnicodeFormatterTest.java [hunk 3]
```diff
@@ -204,4 +204,9 @@ void conversionOfOrdinal4th() {
     void conversionOfOrdinal9th() {
         assertEquals("9ᵗʰ", formatter.format("9\\textsuperscript{th}"));
     }
+
+    @Test
+    void formatPreservesNoBreakingSpaces() {
```
Comment (ryan-carpenter) [inline jablib/src/test/java/org/jabref/logic/layout/format/LatexToUnicodeFormatterTest.java]:
```suggestion
    void formatPreservesNoBreakSpaces() {
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -15,6 +15,8 @@ Note that this project **does not** adhere to [Semantic Versioning](https://semv
 
 ### Fixed
 
+- We fixed an issue where LaTeX to Unicode conversion replaced tildes with standard spaces instead of non-breaking spaces. ([#15158](https://github.com/JabRef/jabref/issues/15158))
```
Comment (ryan-carpenter) [inline CHANGELOG.md]:
```suggestion
- We fixed an issue where LaTeX to Unicode conversion replaced tildes with standard spaces instead of no-break spaces. ([#15158](https://github.com/JabRef/jabref/issues/15158))
```

File: jablib/src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java (hunk 1 above)
Comment (ryan-carpenter) [reply jablib/src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java]: The formal unicode character name is "no-break" space (https://www.unicode.org/charts/charindex.html), which I think is a better choice than the alternative name "non-breaking space" unless JabRef has a precedent of using the alternative. Thanks for the fix, @faneeshh.

**Review by calixtus — APPROVED** [review]

**Comment by Siedlerchr** [comment]
[8](https://github.com/JabRef/jabref/actions/runs/22326303791/job/64597571544?pr=15174#step:4:1149)
BibEntryTest > getFieldOrAliasLatexFreeComplexConversionInAlias() FAILED
    org.opentest4j.AssertionFailedError at BibEntryTest.java:243

**Comment by faneeshh** [comment]
> [8](https://github.com/JabRef/jabref/actions/runs/22326303791/job/64597571544?pr=15174#step:4:1149) BibEntryTest > getFieldOrAliasLatexFreeComplexConversionInAlias() FAILED org.opentest4j.AssertionFailedError at BibEntryTest.java:243



> [8](https://github.com/JabRef/jabref/actions/runs/22326303791/job/64597571544?pr=15174#step:4:1149) BibEntryTest > getFieldOrAliasLatexFreeComplexConversionInAlias() FAILED org.opentest4j.AssertionFailedError at BibEntryTest.java:243

I'm guessing it's failing because it expects a standard space where I've now introduced the no-break space(\u00a0)? If that is the case then should I update the test's expected string to match the new Unicode char? Does that sound right

**Comment by faneeshh** [comment]
It looks like my change is introducing no-break spaces into generated previews and DOIs. I can update the tests to expect \u00a0 but I feel like that  might break DOI links if a user copies them. Should i restrict the conversion so it doesn't affect DOIs or should I update the test assertions

**Comment by calixtus** [comment]
Interesting unexpected side effect. Will label as Devcall discussion.

**Comment by subhramit** [comment]
@faneeshh you can work on something else in the meanwhile if you wish to. Feel free to ping if you get "maximum assignment limit reached".

**Comment by faneeshh** [comment]
> @faneeshh you can work on something else in the meanwhile if you wish to. Feel free to ping if you get "maximum assignment limit reached".

I really appreciate the concern. I was lurking around this SLR support improving issue (https://github.com/JabRef/jabref/issues/12642). I ended up reading through the old reverted PR history and I really feel like I have an overview in my mind of how to fix the missing V2 requirements. Now this might be a bit of a long shot but I could try implementing a cleaner foundation for the study.yaml format and see if I can get the migrator and those string based tests right... @koppor Let me know what you think of the same...

**Comment by subhramit** [comment]
> Now this might be a bit of a long shot but I could try implementing a cleaner foundation for the study.yaml format and see if I can get the migrator and those string based tests right... @koppor Let me know what you think of the same...

Please feel free to go ahead, we can comment further once you give a solution sketch or a draft PR.

**Comment by koppor** [comment]
> like that might break DOI links if a user copies them

Will also happen when spaces are introduced. 

I am not sure when DOIs contain ~

If URLs contain ~, they should be kept. 

Maybe follow-up issue?

File: CHANGELOG.md [hunk 5]
```diff
@@ -17,6 +17,7 @@ Note that this project **does not** adhere to [Semantic Versioning](https://semv
 
 ### Fixed
 
+- We fixed an issue where LaTeX to Unicode conversion replaced tildes with standard spaces instead of non-break spaces. ([#15158](https://github.com/JabRef/jabref/issues/15158))
```
Comment (koppor) [inline CHANGELOG.md]:
```suggestion
- We fixed an issue where LaTeX to Unicode conversion replaced tildes with standard spaces instead of non-break spaces. [#15158](https://github.com/JabRef/jabref/issues/15158)
```

**Comment by koppor** [comment]
> and DOIs.

Why? -- https://www.doi.org/resources/DOI_URI_Scheme.pdf does not contain any reference to ~

**Comment by koppor** [comment]
> I'm guessing it's failing because it expects a standard space where I've now introduced the no-break space(\u00a0)? If that is the case then should I update the test's expected string to match the new Unicode char? Does that sound right

Yes, please - then we can easily see it if it s OK. Now, five people in a call are starring at the screen and not understanding anything. 

Please do it ASAP.

**Comment by faneeshh** [comment]
> > and DOIs.
> 
> Why? -- https://www.doi.org/resources/DOI_URI_Scheme.pdf does not contain any reference to ~

I understand that real DOIs don't contain tildes per the DOI Scheme but the test 6 in CitationStyleGeneratorTest (10.1161/circ.108_827022~special) appeared to be a edge case to test unusual characters which is why I got confused probably. Anyways I've updated the tests to use the no-break space.

**Comment by koppor** [comment]
> > > and DOIs.
> > 
> > 
> > Why? -- https://www.doi.org/resources/DOI_URI_Scheme.pdf does not contain any reference to ~
> 
> I understand that real DOIs don't contain tildes per the DOI Scheme but the test 6 in CitationStyleGeneratorTest (10.1161/circ.108_827022~special) appeared to be a edge case to test unusual characters which is why I got confused probably. Anyways I've updated the tests to use the no-break space.

Learning: Think "out of the box"

Go to https://doi.org/

Enter the doi with ~"

image

See

image

--> Test is wrong. When reviewing the test at hand, the reviewers did not see.

**Comment by faneeshh** [comment]
> Learning: Think "out of the box"

Yeah learning something new with every PR I open... really grateful tho
